app/src/main/java/com/github/se/eduverse/model/MediaCacheManager.kt

```diff
@@ -1,21 +1,53 @@
 package com.github.se.eduverse.model
 
 import android.content.Context
+import androidx.annotation.OptIn
+import androidx.media3.common.util.Log
+import androidx.media3.common.util.UnstableApi
 import com.google.gson.Gson
 import java.io.File
 import java.io.FileOutputStream
 import java.io.IOException
 import java.io.InputStream
-import java.net.URL
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import okhttp3.OkHttpClient
+import okhttp3.Request
 
-// Interface pour le téléchargement
 interface Downloader {
   @Throws(IOException::class) fun openStream(url: String): InputStream
 }
 
 class DefaultDownloader : Downloader {
+  private val client =
+      OkHttpClient.Builder()
+          .connectTimeout(10, java.util.concurrent.TimeUnit.SECONDS)
+          .readTimeout(10, java.util.concurrent.TimeUnit.SECONDS)
+          .writeTimeout(10, java.util.concurrent.TimeUnit.SECONDS)
+          .build()
+
+  @OptIn(UnstableApi::class)
   override fun openStream(url: String): InputStream {
-    return URL(url).openStream()
+    Log.d("DOWNLOAD_DEBUG", "Starting download for URL: $url")
+    try {
+      val request = Request.Builder().url(url).build()
+      val response = client.newCall(request).execute()
+
+      if (!response.isSuccessful) {
+        Log.e(
+            "DOWNLOAD_ERROR",
+            "Request failed with code: ${response.code} and message: ${response.message}")
+        throw IOException("Failed to download file: ${response.code}")
+      }
+
+      val contentLength = response.body?.contentLength() ?: 0L
+      Log.d("DOWNLOAD_DEBUG", "Response successful. Content-Length: $contentLength bytes")
+
+      return response.body?.byteStream() ?: throw IOException("Empty response body for URL: $url")
+    } catch (e: Exception) {
+      Log.e("DOWNLOAD_ERROR", "Exception while downloading file: ${e.message}", e)
+      throw e
+    }
```

Comment:
Logs important details, such as the start of the download, response status, and content length. This aids in debugging and monitoring.
Handles unsuccessful responses and exceptions gracefully with proper logging and descriptive error messages.
Ensures the response body and content length are validated, reducing the risk of null pointer exceptions.